mono/metadata/object.c

```diff
@@ -7574,25 +7597,47 @@ mono_utf16_to_utf8 (const gunichar2 *s, gsize slength, MonoError *error)
 		memcpy (as2, as, written);
 		g_free (as);
 		as = as2;
+
+		// FIXME utf8_length is ambiguous here.
+		// For now it is what strlen would report.
+		// A lot of code does not deal correctly with embedded nuls.
 	}
 
 	return as;
 }
 
+/**
+ * mono_utf16_to_utf8:
+ */
+char *
+mono_utf16_to_utf8 (const gunichar2 *s, gsize slength, MonoError *error)
+{
+	gsize utf8_length = 0;
+	return mono_utf16_to_utf8len (s, slength, &utf8_length, error);
+}
+
 char *
 mono_string_to_utf8_checked_internal (MonoString *s, MonoError *error)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 
 	error_init (error);
-
 	if (s == NULL)
 		return NULL;
+	return mono_utf16_to_utf8 (mono_string_chars_internal (s), s->length, error);
+}
 
-	if (!s->length)
-		return g_strdup ("");
+char *
+mono_string_to_utf8len (MonoStringHandle s, gsize *utf8len, MonoError *error)
+{
+	*utf8len = 0;
+	if (MONO_HANDLE_IS_NULL (s))
+		return NULL;
 
-	return mono_utf16_to_utf8 (mono_string_chars_internal (s), s->length, error);
+	uint32_t gchandle = 0;
+	char *utf8 = mono_utf16_to_utf8len (mono_string_handle_pin_chars (s, &gchandle), mono_string_handle_length (s), utf8len, error);
```

Comment:
could be done with `NO_SAFEPOINTS` now.

Reply:
Agreed. I think this is an accidental leftover, having carried this commit along through several PRs.